wasm/jit/jit_arm64.go

```diff
@@ -336,15 +431,122 @@ func (c *arm64Compiler) compileGlobalSet(o *wazeroir.OperationGlobalSet) error {
+	return c.branchInto(o.Target)
+}
+
+// compileBrIf implements compiler.compileBrIf for the arm64 architecture.
+func (c *arm64Compiler) compileBrIf(o *wazeroir.OperationBrIf) error {
+	cond := c.locationStack.pop()
+
+	jmpWithCond := c.newProg()
+	jmpWithCond.To.Type = obj.TYPE_BRANCH
+	if cond.onConditionalRegister() {
+		switch cond.conditionalRegister {
+		case arm64.COND_EQ:
+			jmpWithCond.As = arm64.ABEQ
+		case arm64.COND_NE:
+			jmpWithCond.As = arm64.ABNE
+		case arm64.COND_HS:
+			jmpWithCond.As = arm64.ABHS
+		case arm64.COND_LO:
+			jmpWithCond.As = arm64.ABLO
+		case arm64.COND_MI:
+			jmpWithCond.As = arm64.ABMI
+		case arm64.COND_HI:
+			jmpWithCond.As = arm64.ABHI
+		case arm64.COND_LS:
+			jmpWithCond.As = arm64.ABLS
+		case arm64.COND_GE:
+			jmpWithCond.As = arm64.ABGE
+		case arm64.COND_LT:
+			jmpWithCond.As = arm64.ABLT
+		case arm64.COND_GT:
+			jmpWithCond.As = arm64.ABGT
+		case arm64.COND_LE:
+			jmpWithCond.As = arm64.ABLE
+		default:
+			// This means that we use the cond.conditionalRegister somewhere in this file,
+			// but not covered in switch ^. That shouldn't happen.
+			return fmt.Errorf("unsupported condition for br_if: %v", cond.conditionalRegister)
+		}
+	} else {
+		// If the value is not on the conditional register, we compare the value with the zero register,
+		// and then do the conditional jump if the value does't equal zero.
+		if err := c.ensureOnGeneralPurposeRegister(cond); err != nil {
+			return err
+		}
+		// Compare the value with zero register. Note that the value is ensured to be i32 by function validation phase,
+		// so we use CMPW (32-bit compare) here.
+		c.applyTwoRegistersToNoneInstruction(arm64.ACMPW, cond.register, zeroRegister)
+		jmpWithCond.As = arm64.ABNE
+
+		c.markRegisterUnused(cond.register)
+	}
+
+	c.addInstruction(jmpWithCond)
+
+	// Emit the code for branching into else branch.
+	// We save and clone the location stack because we might end up modifying it inside of branchInto,
+	// and we have to avoid affecting the code generation for Then branch afterwards.
+	saved := c.locationStack
+	c.replaceLocationStack(saved.clone())
+	if err := c.emitDropRange(o.Else.ToDrop); err != nil {
+		return err
+	}
+	c.branchInto(o.Else.Target)
+
+	// Now ready to emit the code for branching into then branch.
+	// Retrieve the original value location stack so that the code below wont'be affected by the Else branch ^^.
+	c.replaceLocationStack(saved)
+	// We jump here from the original conditinoal jump (jmpWithCond).
+	c.setJmpTargetOnNext(jmpWithCond)
+	if err := c.emitDropRange(o.Then.ToDrop); err != nil {
+		return err
+	}
+	c.branchInto(o.Then.Target)
+	return nil
+}
+
+func (c *arm64Compiler) branchInto(target *wazeroir.BranchTarget) error {
+	if target.IsReturnTarget() {
 		return c.returnFunction()
 	} else {
-		return fmt.Errorf("TODO: only return target is available on arm64")
+		labelKey := target.String()
+		targetLabel := c.label(labelKey)
+		if targetLabel.callers > 1 {
```

Comment:
zero possible and allowed in wasm:

```
(block
  (br 0)
  (block i32.const 1111)
)
```

This example the label corresponding to `(block i32.const 1111)` is never be reached _at runtime_ because `br 0` exits the function before we reach there